### PR TITLE
[FIX] Average Pbar Metrics

### LINKED ISSUE #4395: loss from progress bar appears to be sum of loss across all GPUs in Lightning 1.0.3
## 🐛 Bug

The loss from progress bar during the training appears to be sum of loss across of all GPUs

### To Reproduce

One can pick any model and dataset and varies the number of GPUs for training. In my example, I have roughly 0.7 loss for 1 GPUs, 1.4 loss for 2 GPUs and 0.28 loss for 4 GPUs **for the first few training batches**, indicating loss is roughly the **sum** instead of **mean** of loss across all GPUs.

I used the standard from the doc:

```
def train_step(self, batch, batch_idx):
    loss = self.forward(batch)
   self.log('loss/train', loss, on_step=True, on_epoch=False, sync_dist=True, sync_dist_op='mean')

```

Note: the loss logged in tensorboard appears to be correct (mean of loss)

### Expected behavior
The loss from progress bar during the training should be **mean** of loss across of all GPUs

### Environment

**Note**: `Bugs with code` are solved faster ! `Colab Notebook` should be made `public` !

* `IDE`: Please, use our python [bug_report_model.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py
) template.

* `Colab Notebook`: Please copy and paste the output from our [environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py) (or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

 - PyTorch Version (e.g., 1.0):  1.4
 - OS (e.g., Linux):  Linux
 - How you installed PyTorch (`conda`, `pip`, source):  pip
 - Build command you used (if compiling from source):
 - Python version:  3.7
 - CUDA/cuDNN version:  10.1
 - GPU models and configuration:
 - Any other relevant information:   PyTorch Lightning 1.0.3

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

PB: loss from progress bar appears to be sum of loss across all GPUs in Lightning 1.0.3

The sync_dist being an in-place operation, it was reducing across gpu and then it was changing behaviour.

Solution: Clone the value if sync_dist.

TODO: Find a way to make the `sync_dist` only once. 

Closes #4395

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.    
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/trainer/legacy_deprecate_flow_log_tests/test_trainer_steps_scalar_return.py [hunk 1]
```diff
@@ -190,3 +192,33 @@ def test_train_step_epoch_end_scalar(tmpdir):
...
+@pytest.mark.skipif(torch.cuda.device_count() < 2, reason="test requires multi-GPU machine")
+def test_dpp_reduce_mean_pbar(tmpdir):
+    class ExtentedModel(BoringModel):
+
+        logged = []
+
+        def training_step(self, batch, batch_idx):
+            output = self.layer(batch)
+            loss = self.loss(batch, output)
+            loss /= loss.clone().detach()
```
Comment (s-rog) [inline tests/trainer/legacy_deprecate_flow_log_tests/test_trainer_steps_scalar_return.py:206]:
```suggestion
            loss /= loss.clone().detach()
            self.log('self_log', loss, sync_dist=True)
```
Should reproduce the bug now :]

File: pytorch_lightning/core/step_result.py [hunk 2]
```diff
@@ -129,7 +129,14 @@ def log(
     ):
         # no metrics should be logged with graphs
         if not enable_graph and isinstance(value, torch.Tensor):
-            value = value.detach()
+            # TODO: Find a way to make the reduction only once, so we don't need to clone the
+            is_dist = torch.distributed.is_available() and torch.distributed.is_initialized()
+            if sync_dist and is_dist:
+                # make sure we don't make an in-place operation
+                # .detach().clone() and not .clone().detach() which will copy the entire graph.
+                value = value.detach().clone()
```
Comment (s-rog) [inline pytorch_lightning/core/step_result.py:137]: What's preventing the clone() from being in the `if sync_dist` block directly below?

Comment (tchaton) [reply pytorch_lightning/core/step_result.py:137]: Great catch ! Nothing really :)

Comment (s-rog) [reply pytorch_lightning/core/step_result.py:137]: Gotcha! and is that what you meant by this? I'm not too familiar with the training loop after all these refactors haha
> Find a way to make the sync_dist only once.

**Comment by SeanNaren** [comment]
Looks good! as mentioned I don't think its super priority to worry about a duplicate reduce call on a 1 item tensor, and for explicitness it's better to do it as a part of this func!

**Review by SeanNaren — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]
LGTM

File: pytorch_lightning/core/step_result.py [hunk 3]
```diff
@@ -134,6 +134,9 @@ def log(
         # sync across workers when using distributed training
         sync_fn = sync_fn or sync_ddp_if_available
         if sync_dist and isinstance(value, (torch.Tensor, numbers.Number)):
+            is_dist_initialized = torch.distributed.is_available() and torch.distributed.is_initialized()
```
Comment (Borda) [inline pytorch_lightning/core/step_result.py:137]: is here the available needed?

File: tests/trainer/legacy_deprecate_flow_log_tests/test_trainer_steps_scalar_return.py [hunk 4]
```diff
@@ -190,3 +192,49 @@ def test_train_step_epoch_end_scalar(tmpdir):
...
+
+    trainer = Trainer(
+        max_epochs=1,
+        default_root_dir=os.getcwd(),
+        limit_train_batches=10,
+        limit_test_batches=2,
+        limit_val_batches=2,
+        distributed_backend=distributed_backend,
+        gpus=2,
+        precision=32)
```
Comment (Borda) [inline tests/trainer/legacy_deprecate_flow_log_tests/test_trainer_steps_scalar_return.py:227]:
```suggestion
        precision=32,
    )
```

File: tests/trainer/legacy_deprecate_flow_log_tests/test_trainer_steps_scalar_return.py [hunk 5]
```diff
@@ -190,3 +192,49 @@ def test_train_step_epoch_end_scalar(tmpdir):
...
+
+@pytest.mark.skipif(torch.cuda.device_count() < 2, reason="test requires multi-GPU machine")
+def test_dpp_reduce_mean_pbar(tmpdir):
+    os.environ['PL_DEV_DEBUG'] = '1'
+
+    model = DPPReduceMeanPbarModel()
+    model.training_step_end = None
+    model.training_epoch_end = None
+
+    distributed_backend = "ddp_spawn"
```
Comment (Borda) [inline tests/trainer/legacy_deprecate_flow_log_tests/test_trainer_steps_scalar_return.py:217]: shall it be parameter/parameterize func even for now with a single option?

File: tests/trainer/legacy_deprecate_flow_log_tests/test_trainer_steps_scalar_return.py [hunk 6]
```diff
@@ -190,3 +192,49 @@ def test_train_step_epoch_end_scalar(tmpdir):
...
+
+    # TODO: Move this test to DDP. pbar_added_metrics is empty with ddp_spawn for some reasons
+
+    pbar_added_metrics = trainer.dev_debugger.pbar_added_metrics
+    is_in = False
+    for pbar_metrics in pbar_added_metrics:
+        if 'loss_2' in pbar_metrics:
+            is_in = True
+            assert pbar_metrics["loss_2"].item() == 1
+    if distributed_backend == "ddp":
```
Comment (Borda) [inline tests/trainer/legacy_deprecate_flow_log_tests/test_trainer_steps_scalar_return.py:239]: so this one teste with `distributed_backend = "ddp_spawn"`?

Comment (tchaton) [reply tests/trainer/legacy_deprecate_flow_log_tests/test_trainer_steps_scalar_return.py:239]: The test doesn't really work in "ddp_spawn" for some reasons. I added a TODO to convert it into a DDP one.

File: pytorch_lightning/core/step_result.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/core/step_result.py:137]: Yes.

Comment (tchaton) [reply pytorch_lightning/core/step_result.py:137]: The synchronisation won't happen if `is_dist_initialized` is False, and therefore there is no need to clone the variable.

**Comment by seannz** [comment]
This fix seems to be broken in the latest master (commit 0e19d16ca6f19e13bc7721f48d16ab6584fe1275).
